src/v/crypto/key.cc

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "key.h"
+
+#include "crypto/crypto.h"
+#include "crypto/exceptions.h"
+#include "crypto/internal.h"
+#include "crypto/ssl_utils.h"
+#include "crypto/types.h"
+
+#include <absl/container/flat_hash_set.h>
+#include <openssl/bio.h>
+#include <openssl/bn.h>
+#include <openssl/evp.h>
+#include <openssl/param_build.h>
+#include <openssl/pem.h>
+
+namespace crypto {
+
+namespace internal {
+EVP_PKEY_ptr load_public_pem_key(bytes_view key) {
+    BIO_ptr key_bio(BIO_new_mem_buf(key.data(), static_cast<int>(key.size())));
+    EVP_PKEY_ptr pkey(
+      PEM_read_bio_PUBKEY(key_bio.get(), nullptr, nullptr, nullptr));
+    if (!pkey) {
+        throw ossl_error("Failed to load PEM public key");
+    }
+
+    return pkey;
+}
+
+EVP_PKEY_ptr load_private_pem_key(bytes_view key) {
+    BIO_ptr key_bio(BIO_new_mem_buf(key.data(), static_cast<int>(key.size())));
+    EVP_PKEY_ptr pkey(
+      PEM_read_bio_PrivateKey(key_bio.get(), nullptr, nullptr, nullptr));
+    if (!pkey) {
+        throw ossl_error("Failed to load PEM private key");
+    }
+
+    return pkey;
+}
```

Comment:
With `BIO_new_mem_buf` the onus is on the user to keep the backing of `key` alive until the key is destroyed. I think this API is easy to misuse.

Some suggestions:
1. The caller should transfer ownership of `bytes` at the call-site.
2. The view should be copied to a buffer with a lifetime that's managed

Reply:
Sorry when you said `key` are you referring to `pkey`?  The contents of `key` is a PEM encoded public key which `PEM_read_bio_PUBKEY` will parse and then copy parts of that structure into `BN`s within the RSA key object.  I think `key` can be destroyed after `PEM_read_bio_PUBKEY` is successful and the buffer of bytes does not have to live along with the `EVP_PKEY`.

Reply:
No, I was referring to `key`.

I'll have to trust you because I can't find any documentation and I've spent 20 minutes chasing the implementation and am none the wiser.

Reply:
OK, I wrote a test and it didn't crash.